File: android/Podcast/app/src/test/java/io/f7z/podcast/DomainFrameFixtures.kt
@file:OptIn(ExperimentalUnsignedTypes::class)

package io.f7z.podcast

import com.google.flatbuffers.FlatBufferBuilder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import org.nmp.android.TypedProjectionEnvelope

object DomainFrameFixtures {
    fun decodeDomainFrames(raw: String?): PodcastDomainFrames? {
        raw ?: return null
        return runCatching {
            val outerObj = SnapshotCodec.json.parseToJsonElement(raw) as? JsonObject ?: return null
            val tag = outerObj["t"]?.jsonPrimitive?.contentOrNull ?: return null
            if (tag != "snapshot") return null
            val projections = outerObj["v"]?.jsonObject
                ?.get("projections")
                ?.jsonObject
                ?: return null

            val envelopes = projections.mapNotNull { (key, value) ->
                if (!key.startsWith("podcast.")) return@mapNotNull null
                val rev = runCatching {
                    value.jsonObject["rev"]?.jsonPrimitive?.longOrNull?.toULong()
                }.getOrNull() ?: 1uL
                TypedProjectionEnvelope(
                    key = key,
                    schemaId = key,
                    schemaVersion = 1u,
                    fileIdentifier = "PJPR",
                    payload = encodePjpr(value.toString()),
                    projectionRev = rev,
                    state = 0u.toUByte(),
                )
            }
            val frames = TypedProjectionDomainCodec.decodeDomainFrames(envelopes)
            val resolvedProfiles = SnapshotCodec.decodeResolvedProfiles(raw)
            frames
                ?.copy(resolvedProfiles = frames.resolvedProfiles + resolvedProfiles)
                ?: PodcastDomainFrames(resolvedProfiles = resolvedProfiles).takeIf { it.hasAnyDomain }
        }.getOrNull()
    }

    private fun encodePjpr(json: String): ByteArray {
        val fbb = FlatBufferBuilder()
        val jsonOffset = fbb.createString(json)
        fbb.startTable(2)
        fbb.addInt(0, 1, 0)
        fbb.addOffset(1, jsonOffset, 0)
        val root = fbb.endTable()
        fbb.required(root, 6)
        fbb.finish(root, "PJPR")
        return fbb.sizedByteArray()
    }
}

File: AppTests/Sources/DomainFrameFixtures.swift
import FlatBuffers
import Foundation
@testable import Podcastr

enum DomainFrameFixtures {
    static func decode(from data: Data) -> PodcastDomainFrames? {
        guard
            let raw = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let value = raw["v"] as? [String: Any],
            let projections = value["projections"] as? [String: Any]
        else { return nil }

        let envelopes: [TypedProjectionEnvelope] = projections.compactMap { key, body in
            guard key.starts(with: "podcast.") else { return nil }
            guard JSONSerialization.isValidJSONObject(body),
                  let jsonData = try? JSONSerialization.data(withJSONObject: body),
                  let json = String(data: jsonData, encoding: .utf8)
            else { return nil }

            return TypedProjectionEnvelope(
                key: key,
                schemaId: key,
                schemaVersion: 1,
                fileIdentifier: "PJPR",
                payload: encodePjpr(json: json),
                projectionRev: projectionRev(body),
                state: .changed
            )
        }

        var frames = PodcastDomainFrames.decode(from: envelopes) ?? PodcastDomainFrames()
        frames.resolvedProfiles = PodcastDomainFrames.decodeResolvedProfiles(from: data)
        return frames.hasAnyDomain ? frames : nil
    }

    private static func encodePjpr(json: String) -> Data {
        var fbb = FlatBufferBuilder()
        let jsonOffset = fbb.create(string: json)
        let root = podcastr_projection_PodcastProjectionJsonFrame
            .createPodcastProjectionJsonFrame(&fbb, schemaVersion: 1, jsonOffset: jsonOffset)
        podcastr_projection_PodcastProjectionJsonFrame.finish(&fbb, end: root)
        return Data(fbb.sizedByteArray)
    }

    private static func projectionRev(_ body: Any) -> UInt64 {
        guard let dict = body as? [String: Any] else { return 1 }
        if let rev = dict["rev"] as? UInt64 { return rev }
        if let rev = dict["rev"] as? Int { return UInt64(max(rev, 0)) }
        if let rev = dict["rev"] as? NSNumber { return rev.uint64Value }
        return 1
    }
}
